core/src/main/java/org/apache/iceberg/RemoveSnapshots.java

```diff
@@ -163,21 +188,100 @@ private TableMetadata internalApply() {
     this.base = ops.refresh();
 
     Set<Long> idsToRetain = Sets.newHashSet();
-    List<Long> ancestorIds = SnapshotUtil.ancestorIds(base.currentSnapshot(), base::snapshot);
-    if (minNumSnapshots >= ancestorIds.size()) {
+    // Identify refs that should be removed
+    Set<String> expiredRefs = computeExpiredRefs(base.refs());
+    Set<Long> idsOfRetainedRefs = base.refs().entrySet().stream()
+        .filter(ref -> !expiredRefs.contains(ref.getKey()))
+        .map(refEntry -> refEntry.getValue().snapshotId())
+        .collect(Collectors.toSet());
+
+    for (Long idToRemove : idsToRemove) {
+      Preconditions.checkArgument(!idsOfRetainedRefs.contains(idToRemove),
+          "Cannot expire %s. There is a reference to it", idToRemove);
+    }
+
+    idsToRetain.addAll(idsOfRetainedRefs);
+
+    List<Map.Entry<String, SnapshotRef>> branches = base.refs().entrySet().stream()
+        .filter(refEntry -> refEntry.getValue().isBranch())
+        .collect(Collectors.toList());
+
+    Set<Long> branchSnapshotsToRetain = computeAllBranchSnapshotsToRetain(branches);
+    idsToRetain.addAll(branchSnapshotsToRetain);
+    TableMetadata.Builder updatedMetaBuilder = TableMetadata.buildFrom(base);
+    idsToRemove.addAll(base.snapshots().stream().map(Snapshot::snapshotId)
+        .filter(snapshot -> !idsToRetain.contains(snapshot)).collect(Collectors.toSet()));
+    updatedMetaBuilder.removeSnapshots(idsToRemove);
+    expiredRefs.forEach(updatedMetaBuilder::removeRef);
+    TableMetadata updatedMetadata = updatedMetaBuilder.build();
+    List<Snapshot> updateSnapshots = updatedMetadata.snapshots();
+    List<Snapshot> baseSnapshots = base.snapshots();
+    boolean updated = updateSnapshots.size() != baseSnapshots.size() || expiredRefs.size() > 0;
+    return updated ? updatedMetadata : base;
+  }
+
+  private Set<Long> computeAllBranchSnapshotsToRetain(List<Map.Entry<String, SnapshotRef>> branches) {
```

Comment:
It's a bit strange to pass `List<Map.Entry<String, SnapshotRef>>` here, but `Map<String, SnapshotRef>` to other methods. I'd probably just move branch filtering into this, since you're already looping over the map entries.

Reply:
Yes, I somehow missed cleaning this up. Updated to just loop over the map entries directly and if it's a branch do the desired computation